File: ObjC-JSON-API/ObjC-JSON-API/ViewController.m
//
//  ViewController.m
//  ObjC-JSON-API
//
//  Created by Domenico Vacchiano on 07/11/15.
//  Copyright © 2015 DomenicoVacchiano. All rights reserved.
//

#import "ViewController.h"

#define kTagLabelName  1
#define kTagLabelUrl  2

@interface ViewController ()
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (nonatomic,strong)NSArray*ds;
@end

@implementation ViewController

#pragma mark View life cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    self.tableView.dataSource=self;
    self.activityIndicator.hidden=true;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}
#pragma mark IBAction(s)
-(IBAction)sendWebRequest:(id)sender{
    
    //sends a request
    [L3SDKJWARequest sendTo:@"https://api.github.com/users/alchimya/repos"
            withMethod:HTTPRequestMethodGet
             andParams:nil andDelegate:self];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.activityIndicator.hidden=false;
        [self.activityIndicator startAnimating];
    });
}
#pragma mark JWA callbacks

-(void)L3SDKJWARequestDelegate_Response:(id)response{

    //web data response presentation
    if ([response isKindOfClass:[NSDictionary class]]) {
        NSLog(@"jsons response has a dictionary format");
        
    }else if ([response isKindOfClass:[NSArray class]]) {
        dispatch_async(dispatch_get_main_queue(),^{
            self.ds=response;
            [self.tableView reloadData];
            self.activityIndicator.hidden=true;
            [self.activityIndicator stopAnimating];
        });

    }
    
    //NSLog(@"%@",response);
}
-(void)L3SDKJWARequestDelegate_Error:(NSError*)error{
    NSLog(@"%@",[error localizedDescription]);
    self.activityIndicator.hidden=true;
    [self.activityIndicator stopAnimating];
}

#pragma mark TableView Datasource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.ds.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    
    UITableViewCell*cell=[tableView dequeueReusableCellWithIdentifier:@"cell"];
    UILabel*lName=(UILabel*)[cell viewWithTag:kTagLabelName];
    UILabel*lUrl=(UILabel*)[cell viewWithTag:kTagLabelUrl];
    
    NSDictionary*repo=self.ds[indexPath.row];
    if (repo) {
        lName.text=[repo objectForKey:@"name"];
        lUrl.text=[repo objectForKey:@"url"];
    }
    
    return cell;
}

@end

File: Swift-JSON-API/Swift-JSON-API/ViewController.swift
//
//  ViewController.swift
//  Swift-JSON-API
//
//  Created by Domenico Vacchiano on 07/11/15.
//  Copyright © 2015 DomenicoVacchiano. All rights reserved.
//

import UIKit

class ViewController: UIViewController,L3SDKJWARequestDelegate, UITableViewDataSource {
   
    let kTagLabelName = 1
    let kTagLabelUrl = 2
    
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    
    var ds:NSArray = []
    //MARK View life cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        self.tableView.dataSource=self
        activityIndicator.hidden=true;
        
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    //MARK IBAction(s)
    @IBAction func sendWebRequest(sender:AnyObject){
        //send a request
        L3SDKJWARequest.sendTo(
            url: "https://api.github.com/users/alchimya/repos",
            withMethod: L3SDKJWARequest.HTTPRequestMethod.GET,
            andParams: nil,
            andDelegate: self)

        
        dispatch_async(dispatch_get_main_queue(), {
                self.activityIndicator.hidden=false
                [self.activityIndicator .startAnimating()]
            })
    }

    //MARK JWA callbacks
    func L3SDKJWARequestDelegate_Response (response:AnyObject?){
       
        //web data response presentation
        if let _ = response as? NSDictionary {
            //do something with your json dictionary
            print("jsons response has a dictionary format")
        }
        else if let jsonArray = response as? NSArray {
            //do something with your json array
            
            dispatch_async(dispatch_get_main_queue(),{
                self.ds=jsonArray
                [self.tableView .reloadData()]
                self.activityIndicator.hidden=true
                [self.activityIndicator .stopAnimating()]
            })
            /*
            for item in jsonArray {
                let repo = item as! NSDictionary
                print(repo.objectForKey("name"))
                print(repo.objectForKey("url"))
            }
            */
            
        }
       // print(response)
        

        
    }

    func L3SDKJWARequestDelegate_Error(error:NSError){
        print(error)
        self.activityIndicator.hidden=true
        [self.activityIndicator .stopAnimating()]
    }
    
    //MAKR tableview datasource

    internal func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int{
        return self.ds.count;
    }

    internal func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell{
        let cell = tableView.dequeueReusableCellWithIdentifier("cell")
        
        let lName = cell?.viewWithTag(kTagLabelName) as! UILabel
        let lUrl = cell?.viewWithTag(kTagLabelUrl)as! UILabel
        
        let repo=self.ds[indexPath.row]
    
        lName.text=repo.objectForKey("name") as? String
        lUrl.text=repo.objectForKey("url") as? String
        
        return cell!
    }
    
    
}

